calodiffusion/models/sample.py

```diff
@@ -1121,3 +1121,152 @@ def __call__(self, model, start, energy, layers, num_steps, sample_offset, debug
         
         return self.sampler(start, debug=debug, offset=sample_offset)
 
+class PushforwardTraining(Sample): 
+    """Pushforward sampling to be paired with IMM loss - https://arxiv.org/abs/2503.07565"""
+    def __init__(self, config):
+        super().__init__(config)
+        self.epsilon = self.sample_config.get("EPSILON", 0.01)
+        self.noise_low = self.sample_config.get("NOISE_LOW", 0.0)
+        self.noise_high = self.sample_config.get("NOISE_HIGH", 0.994)
+
+        self.t_distribution = torch.distributions.LogNormal(
+            self.sample_config.get("P_MEAN", 0.0),
+            self.sample_config.get("P_STD", 0.1)
+        )
+
+        self.s_distribution = torch.distributions.Uniform(
+            self.noise_low, 
+            self.noise_high
+        )
+
+        self.k_noise_seperation = self.sample_config.get("K", 12)
+        self.noise_schedule = self.sample_config.get("NOISE_SCHEDULE", "fm")
+
+    def compute_noise_offset(self, noise_sampled): 
+        "Convert s -> r, ensuring there is a proper offset between the two sampling parameters"
+
+        u = (self.noise_high - self.noise_low) * (1/2) ** self.k_noise_seperation
+        return (noise_sampled - u).clamp(min=self.noise_low, max=self.noise_high) 
+
+    def noise_ratio_to_time(self, sample): 
+        """Convert the sample of the noise to a timestep """
+
+        if self.noise_schedule == "vp_cosine":
+            t = torch.arctan(sample) / (torch.pi * 0.5) 
```

Comment:
if this is called frequently, it would be *slightly* faster to store a global constant `vp_cosine_factor = 1.0/(torch.pi*0.5)` somewhere and then compute `torch.arctan(sample) * vp_cosine_factor`